Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,200 @@
+.. SPDX-License-Identifier: GPL-2.0-only
+
+===========
+Ancillary Bus
```

Comment:
@ranj063 is name changed discussed with Greg KH? Did I miss the discussion?

Reply:
@paravmellanox the name change was based on our discussions with kernel developers at Intel. This version is based on the feedback provided and is still under review.

Reply:
but to be clear Greg KH was informed in the email thread for the previous series.

Reply:
@plbossart can you please point me to the discussion thread?

Was the requirement of [1] dropped out? if so, can you point me to the thread? Need to ask what changed?

[1] https://patchwork.kernel.org/patch/11280547/#23056771

Reply:
I don't understand your question based on a reference that's more than 6 months old. What requirement are you referring to?

Reply:
@plbossart It doesn't matter its 6 months old conversation. Kernel devel doesn't complete in month or two for fundamental objects. :-)

And also since v5 posted in April/May 2020 of virtbus was as-is, its not really 6 months old.

During the virtbus introduction its primary usage(s) are discussed in the thread [1].
There are two uses of it.
(a) matching service bus to load multiple class drivers (rdma, netdev) i40e, mlx5 etc.
(b) support slice of a device aka sub-function.

This is based on the discussion of [1] continued and solidified in [2] and [3] further.
[1] https://patchwork.kernel.org/patch/11280547/#23056771
[2] https://lore.kernel.org/netdev/20200519092258.GF4655@nanopsycho/
[3] https://marc.info/?l=linux-netdev&m=158555928517777&w=2

Hope it clarifies.

Reply:
the case a) and b) are still in scope, not sure what makes you think otherwise?

Reply:
Parav - Apart from the name change from virtbus to ancillary bus and updates to Documentation bits, the implementation has not really changed from the last submission.
https://lore.kernel.org/linux-rdma/20200520070227.3392100-1-jeffrey.t.kirsher@intel.com/

This sub-function device your referring is just an ancillary
device carved off a PCI function no?
Is there anything in the actual bus implementation that's missing?

Reply:
You can still encapsulate the ancillary_device within domain-specific objects and ops which is available to the ancillary driver upon attachment. So its customizable / extensible from that standpoint.

Reply:
Hi Shiraz,

I do not have specific problem with the name change.
However virtbus, virtdev is lot shorter name in discussions and also in the sysfs etc views.
Keeping just the name change aside,

my concern is with the section "When Should the Ancillary Bus Not Be Used" which says; 
(a) light weight
(b) use beyond matching service

In [1] with Greg KH we agreed to use for (a) and (b) both and updated documentation section limits the usage.
So I believe "When Should the Ancillary Bus Not Be Used" should be removed.

Also "When to use section", need to describe that it should be used when creating subfunctions of a PCI device.

[1] https://patchwork.kernel.org/patch/11280547/#23056985

Reply:
I agree that the creating of PCI subfunctions is in scope for the use of this bus, but it is not inherent in the bus itself, but what is done with the shared object after the matching is performed.  I can add this into the Documentation though as an example of usage.

Reply:
I agree that we should nto limit the usage to PCI devices. In the SOF driver, for example, we're also using for ACPI platforms.

Reply:
@Parav - I think what we are trying to avoid here is pollute the ancillary bus code from any subsystem specific semantics. As long as your SF stuff doesn't do that I think we are ok.

Reply:
@dmertman - Maybe we drop the 'just matching service' verbiage?

Reply:
> 
> 
> I agree that the creating of PCI subfunctions is in scope for the use of this bus, but it is not inherent in the bus itself, but what is done with the shared object after the matching is performed. I can add this into the Documentation though as an example of usage.

Yes and also need to update the section "where it shouldn't be used".
I think 'lightweight' bus should be removed from documentation, code comments etc.
I do not recall any objection from GKH or others name for virtbus.

You might want to as well keep the name as virtbus and virtdev and after posting ask him, if its reasonable to change the name from virtbus to ancillary. Or mention good reason in cover letter for why a name change from virtbus to ancillary.

Reply:
> 
> 
> @parav - I think what we are trying to avoid here is pollute the ancillary bus code from any subsystem specific semantics. As long as your SF stuff doesn't do that I think we are ok.

Hi Shiraz,
I am not fully following your comment.

Given that sf will have more attributes in sysfs. 
(a) irq details
(b) sfnum (sf number, set by the user during device creation time)
(c) resources directory (like pci to tell about what register resources are used)
(d) link to parent pci device -> to tell systemd/udev to name class (rdma, net) devices properly
(e) if needed iommu support in future.

Not sure if you consider these addition as pollution or enhancement.
Initially I wanted separate bus and probed Greg KH for it.
But his response in [2] indicates that its ok for this bus to serve dual purpose.

[2] https://patchwork.kernel.org/patch/11280547/#23056985

If you think otherwise, we should discuss on mailing list first reviving old thread. But I am afraid of shaking GKH again on this as we asked him many times and he settle on this bus... so... :-)

Reply:
if you think some of you and I should get to a phone call, please let me know. We can do tomorrow on 7/17.

Reply:
@paravmellanox you are citing long email threads that are difficult to parse.
For clarify, were you referring to this exchange:

````
> I am working on RFC where PCI device is sliced to create sub-functions.
> Each sub-function/slice is created dynamically by the user.
> User gives sf-number at creation time which will be used for plumbing by
> systemd/udev, devlink ports.

That sounds exactly what is wanted here as well, right?

> This sub-function will have sysfs attributes = sfnumber, irq vectors,
> PCI BAR resource files.
> sfnumber as sysfs file will be used by systemd/udev to have
> deterministic names of netdev and rdma device created on top of
> sub-function's 'struct device'.
> 
> As opposed to that, matching service devices won't have such attributes.
> 
> We stayed away from using mdev bus for such dual purpose in past.

That is good.

> Should we have virtbus that holds 'struct device' created for different
> purpose and have different sysfs attributes? Is it ok?

That's fine to do, I was expecting that to happen.
````

If yes, I don't get your last point or what the problem is with this PR. You can extend the basic ancillary_device structure and embed in whatever context-specific structures you want. But I don't see at all how you could have context-specific sysfs attributes at the basic level. In fact, we don't want ANY context-specific information there.

Reply:
@plbossart,

no problem with PR. I just want documentation "what it should not be used for" to be remove citations that "it should not be used beyond matching service purposes".

Context specific sysfs attributes are easy, it is done through device groups.
struct device *dev->groups.

Additional sysfs attributes will be initialized through groups.
You can refer to rdma_set_device_sysfs_group() in kernel for example.

I didn't understand your point about 'context specific information".
If you mean a device should not pass any device specific in from device side to bus driver side via sysfs, than yes, I agree with you. I was the one who proposed to use container_of() on bus driver side, along with Jason G.
if 'context specific info' is something else, please let me know.

sysfs of ancillary device looks very similar to pci device sysfs.

Reply:
I am familiar with device groups, but that should be handled with your own drivers. 

I don't think we can have default sysfs attributes added by default in the ancillary_device code, that was my point.

In other words, you can use existing kernel capabilities to add your own attributes.

The only point that maybe would be problematic for userspace is that you would need a device 'type' to know what attributes to expect. That maybe something that needs to be added as a default attribute, with an enum that avoid collisions between device types exposes to userspace. e.g if /sys/bus/ancillary/ancillary.1/type stays 'AUDIO' then userspace should not expect any net/rdma capabiltiies.

Reply:
@plbossart exposing device attributes sysfs outside by a vendor driver is discouraged.
even in rdma subsystem we don't allow that anymore. It is potential source of discrepancy.
net class has very strict rule to not allow this either.

vendor cannot add some attributes and ask systemd/udev to follow it. It is the bus property that has to be followed. 
Additionally any ancillary device we expose for a subfunction is inherently multi function device supporting net, rdma, vdpa and more.
So putting type also doesn't make sense.
Exposing sysfs attributes later has some race condition bugs too in termination path. So we get rid of it in rdma.

when device_add() is completed, that means all necessary sysfs files go live. Nothing after that by a vendor driver.

Reply:
then what solution you are suggesting?

We cannot have both a generic light-weight mechanism and burden it with unrelated stuff. The odds of an audio solution sharing any sysfs attributes with rdma are nil

Reply:
> 
> 
> I am familiar with device groups, but that should be handled with your own drivers.
> 
> I don't think we can have default sysfs attributes added by default in the ancillary_device code, that was my point.
> 
> In other words, you can use existing kernel capabilities to add your own attributes.
> 
> The only point that maybe would be problematic for userspace is that you would need a device 'type' to know what attributes to expect. That maybe something that needs to be added as a default attribute, with an enum that avoid collisions between device types exposes to userspace. e.g if /sys/bus/ancillary/ancillary.1/type stays 'AUDIO' then userspace should not expect any net/rdma capabiltiies.

@plbossart 
Also we certainly don't want to expose every ancillary device as subfunction device.

there will be new API such as
ancillary_device_sf_register(ancillary_device, &sf_sysfs_group);

This will do current work + new sysfs.
So base ancillary device is unmodified.

Reply:
who calls this API and when?

Reply:
> 
> 
> then what solution you are suggesting?
> 
> We cannot have both a generic light-weight mechanism and burden it with unrelated stuff. The odds of an audio solution sharing any sysfs attributes with rdma are nil

@plbossart we just crossed each other. :-)

whether a device is audio or rdma, it is decided on the bus driver side. when 'struct device' is created, it only has base sysfs attributes which I described previously and here below.

Below attributes are created optionally.
1. sfnum
2. irqs
3. resources
4. numa node info
5. link to parent pci device

This is done through by having additional APIs than what is already present in PR.

Reply:
> 
> 
> who calls this API and when?



> 
> 
> who calls this API and when?

@plbossart This API is called by the mlx5_core driver when 

user issues devlink port add command.
An example:
$ devlink port add pci/0000.06.00.0/100 flavour pcisf pfnum 1 sfnum 10

content of [2] is more example oriented for subfunction.
Please refer to diagram and it will likely give you more insights.

This is explained in
[1] https://lore.kernel.org/netdev/20200519092258.GF4655@nanopsycho/
and 
[2] https://marc.info/?l=linux-netdev&m=158555928517777&w=2